docs/architecture/adr-008-blocksize-independent-commitment.md

```diff
@@ -0,0 +1,100 @@
+    }
+}
+```
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Negative
+
+1. The amount of subtree roots per commitment is O(sqrt(n)), while n is the number of message shares. The worst case for the number of subtree roots is depicted in the diagram below - an entire block missing one share.  
+  ![Interactive Commitment 2](./assets/complexity.png)  
+The worst case for the current implementation depends on the block size. If it is the worst block size, as in `minimalBlocksize`, it is O(sqrt(n)) as well. On the other hand, if the message is only in one row, then it is O(log(n)).
+Therefore the height of the tree over the subtree roots is in this implementation O(log(sqrt(n))), while n is the number of message shares. In the current implementation, it varies from O(log(sqrt(n))) to O(log(log(n))) depending on the block size.
+
+2. With more subtree roots, the number of merkle proofs will increase. With deeper subtree roots, the size of the merkle proofs will increase. So instead of having a merkle proof from the `DataRoot` to the `originalSubRoot` you will need the merkle proof from `DataRoot` to `2^k` amount of `miniSubRoots` with `k` being the height difference of `originalSubRoot` and `miniSubRoots`. You can optimize this by having only **one** merkle proof from `DataRoot` to `originalSubRoot` and then calculate the `originalSubRoot` from the `k` `miniSubRoots`. Because the merkle proof is created after the block is published, we know the block size and, therefore, if this `originalSubRoot` exists or not.
+
+### Positive
+
+1. A Rollup can include the commitment in the block header *before* posting to Celestia because it is size-independent and does not have to wait for Celestia to confirm the block size. In general, the rollup needs access to this commitment in some form to verify a message inclusion proof guaranteeing data availability, which Rollmint currently does not have access to.
+2. In turn, this would serve as an alternative to [ADR 007](https://github.com/celestiaorg/optimint/blob/main/docs/lazy-adr/adr-007-header-commit-to-shares.md)
+3. Here is one scheme on how a Rollup might use this new commitment in the block header. Let's assume a Rollup that looks like this:  
+  BH1 <-- BH2 <-- BH3 <-- BH4  
+  The Messages that are submitted to Celestia could look like this:  
+  Message 1: B1  
+  The Commitment of B1 is saved into BH1.  
+  Message 2: (BH1+B2)  
+  The Commitment of BH1+B2 is saved into BH2.  
+  Message 3: (BH2+B3)  
+  The Commitment of (BH2+B3) is saved into BH3, and so on.  
+4. Verifying a message inclusion proof could be done with merkle proofs of the subtree roots to the `DataRoot`, recalculating the commitment, and comparing to what's in the rollup block header. If you trust the validator set of Celestia/Rollup (aka Rollmint Light Client) it could be as simple as submitting a proof over the PFD transaction that included the message and then checking if the commitment is the same as in the PFD transaction.
```

Comment:
I think you mentioned why this wasn't the case before but I forgot... Why can't this be done with the current approach? Don't we have to wait for it to be included in a block before we can create a proof for it?

> If you trust the validator set of Celestia/Rollup (aka Rollmint Light Client)

side note, but just to document that while this is currently the case, this will not be after state fraud proofs.

Reply:
Yes. You need to wait for the rollup message to be included in celestia to create a proof that it was included.
 For the verification you need to have committed to a commitment before it was posted to celestia. If you include all possible commitments in the rollup blockheader than verifying the commitment via the pfd transaction is possible as well. Currently we don't save any commitment to the header of the rollmint block.

Reply:
> side note, but just to document that while this is currently the case, this will not be after state fraud proofs.
wdym by that.

Reply:
just that rollmint light clients shouldn't have to trust the validator set after we have fraud proofs for everything